#### Imports

In [3]:
!ls

MURA-v1.1  sample_data


In [4]:
# -*- coding: utf-8 -*-
# !wget -c https://cs.stanford.edu/group/mlgroup/MURA-v1.1.zip
# !unzip MURA-v1.1.zip
# !rm MURA-v1.1.zip
!ls

MURA-v1.1  sample_data


In [5]:
# -*- coding: utf-8 -*-
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from os import listdir
from os.path import isfile, join
from tqdm import tqdm
import keras
pd.options.display.max_colwidth = 100
from keras.applications.mobilenetv2 import MobileNetV2
from keras.applications.densenet import DenseNet169
from keras.applications.inception_v3 import InceptionV3
from keras.applications.resnet50 import ResNet50
from keras.applications.nasnet import NASNetLarge
from keras.preprocessing import image
from keras.applications.mobilenetv2 import preprocess_input
from keras.applications import MobileNet
from keras.callbacks import (EarlyStopping, ModelCheckpoint, ReduceLROnPlateau, TensorBoard)
from keras.layers import Dense, GlobalAveragePooling2D, Input
from keras.metrics import binary_accuracy, binary_crossentropy
from keras.models import Model
from keras.optimizers import SGD, Adam
from keras.preprocessing.image import ImageDataGenerator
from keras.applications.mobilenetv2 import MobileNetV2
from keras.preprocessing import image as k_im_prep
from keras.models import Model
from keras import backend as K
from keras.models import Sequential

Using TensorFlow backend.


**Data Reading**

In [0]:
def paths_n_labels(csv,str_limp):
    #make dataframe
    studies=pd.read_csv(csv, sep=',',header=None)
    #separate study paths and labels of given limp from those of other limps
    limp_studies=studies[studies[0].str.contains(str_limp)==True]
    #make it a numpy
    limp_studies=np.array(limp_studies)
    #limp study folder paths
    limp_paths=[]
    #labels of given limp
    limp_labels=[]
    for i in tqdm( range(limp_studies.shape[0]) ):
        study_path=limp_studies[i][0]
        study_label=limp_studies[i][1]
        study_files = [f for f in listdir(study_path) if isfile(join(study_path, f))]
        for image in study_files:
            limp_paths.append(study_path + image)
            limp_labels.append(study_label)

    limp_paths=np.array(limp_paths)
    limp_labels=np.array(limp_labels)

    return limp_paths,limp_labels

In [0]:
#general function with options for wrist data case, only set wrist_train=True in the case of wrist training data only
#for all other limps and for validation data even that of wrist just pass the paths
#targ_size is image resizing with default(224,224)
#preprocess flag is for using keras preprocessing for images or just resizing
def read_images(paths ,targ_size= (224, 224), wrist_train=False, preprocess=False):
    images=[]
    #load any limp images
    if(not wrist_train):
        for path in tqdm(paths):
            img=k_im_prep.load_img(path, target_size=targ_size )
            if(preprocess):
                img = k_im_prep.img_to_array(img)
                img = np.expand_dims(img, axis=0)
                img = preprocess_input(img)
                images.append(np.array(img)[0])
            else:
              images.append(np.array(img))
    #special case for wrist train corrupted data       
    else:
        #did this because it gave an error at sample  5307 or near it if took all
        sample_e=5307
        sample_s2=5339
        images=[]
        for path in tqdm(paths[:sample_e]):
            img=k_im_prep.load_img(path, target_size=targ_size)
            if(preprocess):
                img = k_im_prep.img_to_array(img)
                img = np.expand_dims(img, axis=0)
                img = preprocess_input(img)
                images.append(np.array(img)[0])
            else:
              images.append(np.array(img))

        #new start
        for path in tqdm(paths[sample_s2:]):
            img=k_im_prep.load_img(path, target_size=targ_size)
            if(preprocess):
                img = k_im_prep.img_to_array(img)
                img = np.expand_dims(img, axis=0)
                img = preprocess_input(img)
                images.append(np.array(img)[0])
            else:
              images.append(np.array(img))

    #making it a numpy array instead of python list
    return (np.array(images))
  

def wrist_labels(labels):
  sample_e=5307
  sample_s2=5339
  return np.hstack( [ labels[:sample_e], labels[sample_s2:] ])


In [0]:
train_studies='MURA-v1.1/train_labeled_studies.csv'
valid_studies='MURA-v1.1/valid_labeled_studies.csv'

In [9]:
#training images paths and labels
#starting with wrist data
train_paths,train_labels=paths_n_labels(train_studies,"WRIST")
#data bias : train
print("0 normal, 1 abnormal")
unique, counts = np.unique(train_labels, return_counts=True)
print(dict(zip(unique, counts)))

100%|██████████| 3460/3460 [00:02<00:00, 1671.33it/s]

0 normal, 1 abnormal
{0: 5769, 1: 3987}


In [0]:
# print("0 normal, 1 abnormal")
# unique, counts = np.unique(valid_labels, return_counts=True)
# print(dict(zip(unique, counts)))

### Model

In [0]:
def make_FT_model(base=1, imagenet=True, freeze_all=True, add_denses=True):
  
  #weights of pretrained model
  if (imagenet==True):
    w='imagenet'
  else:
    w=None
  
  #initializing pretrained model
  if (base==0):
    base_model = MobileNetV2(input_shape= (224, 224, 3),weights=w, include_top=False)
  elif (base == 1):
    base_model = DenseNet169(input_shape= (224, 224, 3),weights=w, include_top=False)
  elif (base == 2):
    base_model = InceptionV3(input_shape= (224, 224, 3),weights=w, include_top=False)
  elif (base == 3):
    base_model = ResNet50(input_shape= (224, 224, 3),weights=w, include_top=False)   
  elif (base == 4):
    base_model = NASNetLarge(input_shape= (224, 224, 3),weights=w, include_top=False)
    
 
  if (freeze_all):
    #freeze layers of densenet
    for layer in base_model.layers:
      layer.trainable= False 
  
  # add a global spatial average pooling layer
  x = base_model.output
  x = GlobalAveragePooling2D()(x)
  
  if(add_denses):
    # let's add a fully-connected layer
    x = Dense(1024, activation='relu')(x)
    x = Dense(512, activation='relu')(x)
    x = Dense(128, activation='relu')(x)
    x = Dense(32, activation='relu')(x)
    # and a logistic layer -- let's say we have 200 classes
    predictions = Dense(1, activation='sigmoid')(x)
    # this is the model we will train
    model = Model(inputs=base_model.input, outputs=predictions)
    
  else:
    # just feature extractor
    model = Model(inputs=base_model.input, output=x)
  
  
  return model

In [0]:
 def images_n_labels(limp): 
  
  print("\nreading studies of "+ limp + "\n")
  print(train_studies)
  train_paths,train_labels=paths_n_labels(train_studies,limp)
  valid_paths,valid_labels=paths_n_labels(valid_studies,limp)
  
  print(train_labels.shape)
  print(valid_labels.shape)
  print("reading "+ limp + " training images")
  if (limp == "WRIST"):
    train_labels=wrist_labels(train_labels) #for wrist reading problem
    train_imgs= read_images(train_paths,preprocess=True, wrist_train=True)
    
  else:
    train_imgs= read_images(train_paths,preprocess=True, wrist_train=False)
  print(train_imgs.shape)  
  print("reading "+ limp + " validation images")
  valid_imgs= read_images(valid_paths,preprocess=True)
  print(valid_imgs.shape)
  
  return train_imgs, train_labels, valid_imgs, valid_labels

### Generalizing over all limps

In [0]:
limps=["SHOULDER", "WRIST", "FINGER", "ELBOW", "HUMERUS", "HAND", "FOREARM"]


#pass to model list of limps because if wanted to train on less
#function outputs a dictionary or dataframe has train and val accuracies for each limp using a chosen model

def evaluate_limps(model_no=1, imagenet=True, freeze_all=False,v=1 , limps=limps, augment=False):
  accuracies={}
  for limp in limps:
    #print("reading "+ limp + " images\n")
    train_imgs, train_labels, valid_imgs, valid_labels= images_n_labels(limp)
    print("making model")
    model=make_FT_model(base= model_no, imagenet=False, freeze_all=False, add_denses=True)
    print("compiling")
    model.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['accuracy'])
    
    ##############################################################
    
    if(augment): #dataaugmentation model fitting
      print("Augmenting Input data")
      model.fit_generator(datagen.flow(train_imgs, train_labels, batch_size=16),
                    steps_per_epoch=len(train_imgs) / 16, epochs=5,use_multiprocessing=True,workers=6)
      
    ###############################################################
    else:
      
      print("fitting")
      model.fit(train_imgs, train_labels, epochs=5, validation_data=(valid_imgs, valid_labels), shuffle=True, verbose=v )
    print("training loss calculations for "+ limp)
    loss_tr, accuracy_tr =model.evaluate(x=train_imgs, y=train_labels, batch_size=128, verbose=v)
    print("calculating validation loss for "+ limp)
    loss_val, accuracy_val =model.evaluate(x=valid_imgs, y=valid_labels, batch_size=128, verbose=v)
    accuracies.update( {limp : [accuracy_tr, accuracy_val]} )
    print(accuracies)
  return accuracies

  
  

In [19]:
R=evaluate_limps(limps=["SHOULDER"],augment=True)


 30%|███       | 847/2821 [00:00<00:00, 8469.36it/s]


reading studies of SHOULDER

MURA-v1.1/train_labeled_studies.csv


  0%|          | 19/8379 [00:00<00:45, 184.73it/s]

(8379,)
(563,)
reading SHOULDER training images


  1%|▏         | 8/563 [00:00<00:07, 77.49it/s]

(8379, 224, 224, 3)
reading SHOULDER validation images


100%|██████████| 563/563 [00:20<00:00, 26.12it/s]


(563, 224, 224, 3)
making model
compiling
Augmenting Input data
Epoch 1/5


/usr/local/lib/python3.6/dist-packages/keras/engine/training_generator.py:47: UserWarning: Using a generator with `use_multiprocessing=True` and multiple workers may duplicate your data. Please consider using the`keras.utils.Sequence class.
  UserWarning('Using a generator with `use_multiprocessing=True`'
/usr/local/lib/python3.6/dist-packages/keras_preprocessing/image.py:1131: UserWarning: This ImageDataGenerator specifies `featurewise_center`, but it hasn't been fit on any training data. Fit it first by calling `.fit(numpy_data)`.
  warnings.warn('This ImageDataGenerator specifies '
/usr/local/lib/python3.6/dist-packages/keras_preprocessing/image.py:1131: UserWarning: This ImageDataGenerator specifies `featurewise_center`, but it hasn't been fit on any training data. Fit it first by calling `.fit(numpy_data)`.
  warnings.warn('This ImageDataGenerator specifies '
/usr/local/lib/python3.6/dist-packages/keras_preprocessing/image.py:1131: UserWarning: This ImageDataGenerator specifies `f

524/523 [==============================] - 395s 754ms/step - loss: 0.6093 - acc: 0.7190
Epoch 2/5
524/523 [==============================] - 349s 666ms/step - loss: 0.5294 - acc: 0.7648
Epoch 3/5
524/523 [==============================] - 347s 663ms/step - loss: 0.5185 - acc: 0.7773
Epoch 4/5
524/523 [==============================] - 349s 665ms/step - loss: 0.4977 - acc: 0.7845
Epoch 5/5
524/523 [==============================] - 346s 660ms/step - loss: 0.4918 - acc: 0.7931
training loss calculations for SHOULDER
8379/8379 [==============================] - 87s 10ms/step
calculating validation loss for SHOULDER
563/563 [==============================] - 5s 10ms/step
{'SHOULDER': [0.48907984227834816, 0.5026642988248998]}


In [0]:
pd.DataFrame(R).head(2)

In [0]:
#may need to make a function to evaluate all limps over one model together as one dataset

** Current point: generalizing functions and code cleaning+ seeing early stopping callback**

All unsolved problems:

* which layers to freeze and which to train + should I train TL before freezing it ?
* data augmentation to generate more data (solved by khaled,still needs little verification)
* recording variation in accuracy after every change to get intuition
* what does outputted loss represent ? how to read the number ?
* normalization step and its effect on accuracy
* should I use 1 or two neurons at output layer ?
* binary crossentropy weights
* justifying parameter use and discovering useful params
* try training with model unfrozen with imagenet and without it
* make a function to record and tabulate outputs
* could we add precision or recall metric ? change accuracy?
* why doesn't it work if removed GlobalAveragePooling line?
* see if want to freeze less layers
* generalize file reading functions
* use better batch size ? increase epochs ?
* training function with preprocessing flag and multiple model comparisons
* grid search like function to tune models and hyperparameters
* early stopping callback and save model to drive for continuing training
* feature concatenation
* #Khaled : the runtime died issue when trying to fit the generated data first before using it in fit_generaton which gives a warning in trainig



**results on wrist data: no preprocessing but resize**


1- Densenet, with imagenet and froze all at rms prop

59% train 55% val

2- Densenet with imagenet and didnt freeze (trained over them)

65% train and 65% validation

3-InceptionV3, with imagenet and froze all at rms prop

59% train 55% val

4-InceptionV3 with imagenet and didnt freeze (trained over them)

82% train and 78% validation



**results on wrist data: with keras preprocessing **

1- Densenet with imagenet and didnt freeze (trained over them)
71% train and 68% validation

epoch=6 mins

2- Densenet WITHOUT imagenet and didnt freeze 

62% train and 59% validation

epoch=5.5 mins

3-InceptionV3 with imagenet and didnt freeze (trained over them)
79% train and 74% validation

epoch=5 mins

4-InceptionV3 WITHOUT imagenet and didnt freeze 
59% train and 57% validation

epoch=5 mins

**results on Shoulder data: no preprocessing 
**

1- Densenet with imagenet and didnt freeze (trained over them) 53.57% train and 51.15% validation

epoch=6 mins

**results DATA AUGMENTATION on Shoulder data:no preprocessing 
**

1- Densenet with imagenet and didnt freeze (trained over them) 48.9% train and 50.2% validation

epoch=6.7 mins


** Conclusions till now **

1- keras preprocessing (nomalization and subtracting mean maybe ) with inception reduces  accuracy both training and validation , but with DenseNet it improved accuracy 

2- Training over imagenet weights without freezing the TL model gives significantly higher accuracy than freezing whole TL and just training denses

3-Training TL without imagenet and no freeze till now (2 Trials) seems to get worse results than with imagenet and training over it

**The  Generator**


TODO #K  : the runtime died issue when trying to fit the generated data first before using it in fit_generator

In [0]:
datagen = ImageDataGenerator(featurewise_center=True,
                     featurewise_std_normalization=True,
                     rotation_range=90,
                     width_shift_range=0.1,
                     height_shift_range=0.1,
                     zoom_range=0.2)

# # compute quantities required for featurewise normalization
# # (std, mean, and principal components if ZCA whitening is applied)
# # datagen.fit(wrist_images_tr)


# print(wrist_images_tr.shape)
# print(wrist_labels_tr.shape)

In [0]:
# model=make_FT_model(base=1, imagenet=True, freeze_all=False, add_denses=True)
# #m.summary()
# #Choose adam or RMSprop ?!
# # compile the model (should be done *after* setting layers to non-trainable)
# model.compile(optimizer='rmsprop',
#               loss='binary_crossentropy',
#               metrics=['accuracy'])

# # fits the model on batches with real-time data augmentation:
# model.fit_generator(datagen.flow(wrist_images_tr, wrist_labels_tr, batch_size=16),
#                     steps_per_epoch=len(train_imgs) / 16, epochs=10,use_multiprocessing=True,workers=6)

# # model.fit(wrist_images_tr, wrist_labels_tr, epochs=5, validation_data=(wrist_images_val, wrist_labels_val), shuffle=True, verbose=1 )

**Old wrist only  evaluation method for data generation purposes**

In [0]:
model=make_FT_model(base=2, imagenet=False, freeze_all=False, add_denses=True)
#m.summary()
#Choose adam or RMSprop ?!
# compile the model (should be done *after* setting layers to non-trainable)
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])
'''
IF USING OTHER DATA THAN WRIST CHANGE w_labels to train_labels
'''
model.fit(train_imgs, w_labels, epochs=5, validation_data=(valid_imgs, valid_labels), shuffle=True, verbose=1 )
loss,accuracy=model.evaluate(x=train_imgs, y=w_labels, batch_size=128, verbose=1)
print("train accuracy before data augmenation",accuracy)
loss,accuracy=model.evaluate(x=valid_imgs, y=valid_labels, batch_size=128, verbose=1)
print("validation accuracy before data augmenation ",accuracy)

In [0]:
model=make_FT_model(base=2, imagenet=False, freeze_all=False, add_denses=True)
#m.summary()
#Choose adam or RMSprop ?!
# compile the model (should be done *after* setting layers to non-trainable)
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])
'''
IF USING OTHER DATA THAN WRIST CHANGE w_labels to train_labels
'''
model.fit(train_imgs, w_labels, epochs=5, validation_data=(valid_imgs, valid_labels), shuffle=True, verbose=1 )
loss,accuracy=model.evaluate(x=train_imgs, y=w_labels, batch_size=128, verbose=1)
print("train accuracy before data augmenation",accuracy)
loss,accuracy=model.evaluate(x=valid_imgs, y=valid_labels, batch_size=128, verbose=1)
print("validation accuracy before data augmenation ",accuracy)